metasmoke.py

```diff
@@ -32,16 +32,103 @@
+
+    class AutoSwitch:
+        """ Automatically switch metasmoke status """
+        # Last change was on 20 May 2020.
+        MAX_FAILURES = 10  # 10 failures == ms down
+        MAX_SUCCESSES = 1  # 1 success == ms up
+        counter = 0
+        auto = True
+        rw_lock = threading.Lock()
+
+        @staticmethod
+        def to_off():
+            """ Road to switching metasmoke down """
+            with Metasmoke.AutoSwitch.rw_lock:
+                if Metasmoke.AutoSwitch.counter < 0:
+                    # Consecutive counter. Switch sign.
+                    Metasmoke.AutoSwitch.counter = 0
+                Metasmoke.AutoSwitch.counter += 1
+                current_counter = Metasmoke.AutoSwitch.counter
+                current_auto = Metasmoke.AutoSwitch.auto
+            # MAX_FAILURES is constant so no lock.
+            if current_counter > Metasmoke.AutoSwitch.MAX_FAILURES and\
+               GlobalVars.MSStatus.is_up() and current_auto:
+                log("warning", "Last {} connections to metasmoke failed".format(Metasmoke.AutoSwitch.counter) +
+                               " Setting metasmoke status to down.")
+                chatcommunicate.tell_rooms_with("debug", "**Warning**. {}:".format(GlobalVars.location) +
+                                                         " Last {} connection to metasmoke".format(current_counter) +
+                                                         " failed. Setting metasmoke status to **down**.")
+                Metasmoke.ms_down()
+
+        @staticmethod
+        def to_on():
+            """ Road to switching metasmoke up """
+            with Metasmoke.AutoSwitch.rw_lock:
+                if Metasmoke.AutoSwitch.counter > 0:
+                    # Consecutive counter. Switch sign.
+                    Metasmoke.AutoSwitch.counter = 0
+                Metasmoke.AutoSwitch.counter -= 1
+                # Negative values for success
+                current_counter = -Metasmoke.AutoSwitch.counter
+                current_auto = Metasmoke.AutoSwitch.auto
+            # MAX_SUCCESSES is constant so no lock.
+            if current_counter > Metasmoke.AutoSwitch.MAX_SUCCESSES and\
+               GlobalVars.MSStatus.is_down() and current_auto:
+                # Why use warning? Because some action may be needed if people don't think metasmoke is up.
+                log("warning", "Last {} connections to metasmoke succeeded".format(Metasmoke.AutoSwitch.counter) +
+                               " Setting metasmoke status to up.")
+                chatcommunicate.tell_rooms_with("debug", "**Warning**. {}:".format(GlobalVars.location) +
+                                                         " Last {} connection to metasmoke".format(current_counter) +
+                                                         " succeeded. Setting metasmoke status to **up**.")
+                Metasmoke.ms_up()
+
+        @staticmethod
+        def switch_auto(on):
```

Comment:
This might be better as `permit_auto_switching(on=True)` or `permit_auto_on_off(on=True)`, etc.